PR to Main: Added GPS module, updated TCP and I2C transaction modules

ganeshram93 opened the pull request.

coder137 committed:
Added BT HC05 module

- Added low level read and write functionality
- Updated main.cpp with btserial

ganeshram93 committed:
Ported Heartbeat logic from TI to ST B-L475E-IOT01A
Added heartbeat data values in the buffer for BT transmit
TODO: Need to make it threaded

ganeshram93 committed:
Made the heartbeat sensor logic threaded
Code clean up

ganeshram93 committed:
1. Renamed some functions
2. TODO: Send heartbeat data to WiFi

Signed-off-by: ganeshram2493 <ganeshram2493@gmail.com>

ganeshram93 committed:
1. Interfaced GPS hardware and added the GPS module
2. Handled cases when there is Inactive GPS mode and if GPS hardware stops buffering data - NA will be sent to BT module

ganeshram93 committed:
1. Removed GPGGA related code
2. Added logic of reading '\n' and processing the line, rather than reading 400 characters at a given time and processing
3. Renamed class, namesspaces, and some function names

ganeshram93 committed:
Updated main.cpp after resolving merge conflicts due to merging of bt branch in main

ganeshram93 committed:
1. Added comments for Interfaces for each sensor
2. Added default coordinate values (0.0, 0.0) that will be set when GPS module fails to function

ganeshram93 committed:
Merge branch 'heartbeat_bt_branch' of github.com:Connected-Healthcare/b_l475e_iot01a_sensors into temp_combined_sensors_br

Integrated internal sensors, GPS (3.3v) and HB (I2C3) and verified
functionality.

TODO:
1. Integrate internal sensors, GPS (3.3v), HB (I2C3), CO sensors and
verify functionality.
2. Integrate internal sensors, GPS (3.3v), HB (I2C3), CO sensors and
connect from BL-475 ST board to TI board via same I2C3 (multiplexing) to
check passing of all sensor data.
3. Clean up SGP30 code in main.c

Signed-off-by: ganeshram2493 <ganeshram2493@gmail.com>

ganeshram93 committed:
Changed (double) to (float) for GPS coordinates

ganeshram93 committed:
Re-formatted user/slave/i2c_slave.cpp file

ganeshram93 committed:
PR to main:
Changes made:
1. Added GPS module
2. Modified variable scope for Heartbeat sensor
3. Commented BT code for now. Will check once the hardware is available
4. Removed SGP30 high-level code (drivers still exist)
5. Handled I2C requests for Heartbeat and GPS
6. Updated TCP server python script by removing SGP30 and adding HB and
GPS data handling cases
7. Other minor formatting

Signed-off-by: ganeshram2493 <ganeshram2493@gmail.com>

ganeshram93 committed:
1. Updated this branch by pulling main branch and resolved merge
conflicts in main.cpp, gps.cpp, gps.hpp:
Merge branch 'main' of github.com:Connected-Healthcare/b_l475e_iot01a_sensors into temp_combined_sensors_br
2. Changed format specifier for HB data in internet.cpp

Signed-off-by: ganeshram2493 <ganeshram2493@gmail.com>

coder137 commented on HelloWorld_ST_Sensors/main.cpp:
```diff
@@ -60,50 +60,45 @@
+static spec::CarbonMonoxide co(PA_0, PA_1);                 // UART4
+static gps::adafruit_PA6H gps_obj(PC_4, PC_5, 9600);        // UART3
+static i2c_slave::SlaveCommunication slave(PC_1, PC_0, co); // I2C1
+// static bt::hc05 btserial(PA_2, PA_3, 9600);             // UART2
```
uncomment the btserial. If no connection occurs there shouldn't be a problem.

ganeshram93 replied:
Done. Updated in new PR.

coder137 commented on HelloWorld_ST_Sensors/main.cpp:
```diff
@@ -60,50 +60,45 @@
-  strcpy(btserial_data, line);
-  is_recv = true;
-}
+// void btserial_process(const char *line)
```
Uncomment

coder137 commented on HelloWorld_ST_Sensors/main.cpp:
```diff
@@ -60,50 +60,45 @@
-  sgp30.start();
   slave.init_thread();
-  btserial.register_process_func(btserial_process);
+  // btserial.register_process_func(btserial_process);
```
Uncomment

coder137 commented on HelloWorld_ST_Sensors/main.cpp:
```diff
@@ -60,50 +60,45 @@
+static gps::adafruit_PA6H gps_obj(PC_4, PC_5, 9600);        // UART3
+static i2c_slave::SlaveCommunication slave(PC_1, PC_0, co); // I2C1
+// static bt::hc05 btserial(PA_2, PA_3, 9600);             // UART2
+static hb_sensor::hb_sensor_class hb_obj(PB_9, PB_8); // I2C3
```
Improve naming
For example: heartbeat::<device_name>

ganeshram93 replied:
Done. Updated in new PR.

coder137 commented on HelloWorld_ST_Sensors/main.cpp:
```diff
@@ -116,72 +111,75 @@ int main()
   debugPrintf("\r\n--- Starting new run ---\r\n\r\n");
   ThisThread::sleep_for(1000);
-  char buffer[200] = {0};
+  // char buffer[200] = {0};
```
Uncomment

ganeshram93 replied:
Done. Updated in new PR.

coder137 commented on HelloWorld_ST_Sensors/user/heartbeat/heartbeat_logic.cpp:
```diff
@@ -0,0 +1,491 @@
+#include "heartbeat_logic.hpp"
```
Why hearttbeat logic? Improve naming for header files as well

ganeshram93 replied:
Done. Updated in new PR.

coder137 commented on HelloWorld_ST_Sensors/user/heartbeat/heartbeat_logic.hpp:
```diff
@@ -0,0 +1,227 @@
+
+  static struct bioData body;
+
+  class hb_sensor_class
```
hb_sensor_class does not give any meaningful information
Put partnumber instead 
for example: hb_sensor::<partnumber>

ganeshram93 replied:
Done. Updated in new PR.

coder137 commented on HelloWorld_ST_Sensors/user/heartbeat/heartbeat_logic.hpp:
```diff
@@ -0,0 +1,227 @@
+    ENABLE_WHRM_ALGO = 0x02
+  };
+
+  static struct bioData body;
```
Why is this static variable in a header file outside the class?
Make it a part of member variables of the class if being used.

Multiple classes will end up modifying the same struct.

ganeshram93 replied:
Done. Updated in new PR.

coder137 reviewed (changes requested):
Another point is that in some files there is no code change just formatting changes.
Make sure your VSCode clang-format formatting is set to LLVM

ganeshram93 committed:
1. Updated PR comments
2. Renamed Heartbeat class and namespaces
3. Removed certain prints
4. Uncommented BT serial and verified working
5. Moved the heartbeat data variable 'body' inside heartbeat.cpp

Signed-off-by: ganeshram2493 <ganeshram2493@gmail.com>

ganeshram93 reviewed (commented):
Done with all reviews. Updated in new PR.

coder137 commented on HelloWorld_ST_Sensors/user/heartbeat/heartbeat.hpp:
```diff
@@ -0,0 +1,223 @@
+    const uint8_t WRITE_SET_THRESHOLD = 0x01;
+  };
+
+  struct bioData &get_hb_data(void);
```
Make this const ref
`const bioData & get_hb_data()`

If you just do `bioData &get_hb_data()` users can modify the data like this
```cpp
get_hb_data().irled = 0 // their data
```

coder137 replied:
You can also drop the struct keyword. That is a C requirement. In C++ you don't need to use the struct preface

ganeshram93 replied:
Done. Updated in PR.

ganeshram93 committed:
PR2_updated:
1. Made get_hb_data() const
2. Updated paths to header file inclusions, relative to user/

Signed-off-by: ganeshram2493 <ganeshram2493@gmail.com>

coder137 commented on HelloWorld_ST_Sensors/main.cpp:
```diff
@@ -117,71 +112,74 @@ int main()
-    gps_coordinates_data.longitude = 0.0;
+
+    // Heartbeat Sensor
+    const heartbeat::bioData hb_data = heartbeat::get_hb_data();
```
Is the heartbeat giving you data since you changed the API?
It will now be `hb_obj.get_hb_data();`

coder137 commented on HelloWorld_ST_Sensors/user/heartbeat/heartbeat.hpp:
```diff
@@ -0,0 +1,223 @@
+    const uint8_t WRITE_SET_THRESHOLD = 0x01;
+  };
+
+  const bioData &get_hb_data(void);
```
Put the getter inside the class right? Why is this outside?
As we discussed when there are multiple instances of this class how will this function behave?

ganeshram93 replied:
Will resolve in next commit.

coder137 replied:
@ganeshram93 This hasnt been resolved correctly.
Make this inside the class?

coder137 commented:
Also use LLVM format for your code
Since everything shows a diff

ganeshram93 committed:
LLVM formatted the code.

Signed-off-by: ganeshram2493 <ganeshram2493@gmail.com>
